### DISCUSSION TITLE
run.test_prameters in avocado 96.0

### DISCUSSION THREAD

**Discussion by marioaag** [avocado-framework/avocado#5345] [Q&A] [answered]
I have an implementation in Avocado 92.0 where I passed the test parameter using run.test_parameters using the config option at the Runnable generation level and it works as expected but moving to Avocado 96.0 the test parameter get lost and I'm not able to retrieve the values at the test level (avocado instrumented) with self.params.get()
Is there a new parameter that should be used to specify the specific parameters corresponding to each one of the Runnable generated?

Example of the code used:
```
runnable = Runnable('avocado-instrumented',
                                     test_ref,
                                     config={'run.test_parameters': variant_parameters})
```

Example of the Runnable generated:
```
<Runnable kind="avocado-instrumented" uri="/home/mario/git/iLabConfig/test/avocado/testcases/SwitchCounters/SwitchCounters.py:SwitchCounters.analyze_data" config="{'run.test_parameters': {'UUT_NAME': 'ilsw56j', 'INVENTORY_DIR': '/home/mario/git/iLabConfig/uuts/switches', 'THRESHOLD': 5}}" args="()" kwargs="{}" tags="None" dependencies="None"> variant="None"
```

Did it a try with kwargs and the params still getting lost:
```
<Runnable kind="avocado-instrumented" uri="/home/mario/git/iLabConfig/test/avocado/testcases/SwitchCounters/SwitchCounters.py:SwitchCounters.retrieve_data" config="{'run.test_parameters': {'UUT_NAME': 'ilsw56j', 'INVENTORY_DIR': '/home/mario/git/iLabConfig/uuts/switches'}}" args="()" kwargs="{'UUT_NAME': 'ilsw56j', 'INVENTORY_DIR': '/home/mario/git/iLabConfig/uuts/switches'}" tags="None" dependencies="None"> variant="None"

**Comment by marioaag** [comment]
Looking in the code found that the logic to retrieve run.test_parameters is missing in avocado_instrumeted.py in the method _create_params

Something link this:
```
paths = ['/']
        tree_nodes = TreeNode().get_node(paths[0], True)
        tree_nodes.value = dict(runnable.config.get('run.test_parameters',
                                                    []))
```

I did an experiment adding this code and it works

Is the approach changing? asking this due this code is in 92.0 and now it is not.
Let me know your thoughts and maybe I can help to create the pull request to add this code back in the new release

**Reply by clebergnu** [reply]
Hi @marioaag,

Thanks for reporting this.  It looks like a regression indeed, and not something that was changed on purpose since 92.0.

Please go ahead and propose a fix, it will be very much appreciated!

**Reply by marioaag** [reply]
Hi @clebergnu,
I created the issue to track the details #5347
Also created the pull request #5348

It already pass the all the CI checks! is any ETA for the next release?

**Reply by richtja** [reply]
Hi @marioaag, IMO your approach with Runnable creation is not correct. You can't use different config for Runnable and Suite. I mean that every Runnable in one Suite should have the same config. If you want to set the parameters via Job API, you should use `Runnable.variant`.  Sorry for my late reply, I saw that you already prepared some fixes which I think are not needed.

**Reply by marioaag** [reply]
ooooo, I saw the attribute "variant" but I didn't find any example of how to populate it via API
@richtja, could you provide an example of the structure that I need to use to pas the parameters through the variant attribute?
just to confirm, the values passed through the "variant" attribute will be accessible via self.params.get() at the test level, correct?

**Reply by richtja** [reply]
Yes, when you use the "variant" attribute, the value will be accessible via self.params.get(). Unfortunately, when I was creating the example for you, I found out that we have a bug there, so I created issue #5356. Until this will be fixed, you can use 'run.test_parameters' in suite config or some of the avocado varianter plugins, which will create tests based on your parameters for you. But I am not sure it those options suite you.

**Reply by marioaag** [reply]
@richtja Where I can find information about avocado varianter plugins?

**Reply by richtja** [reply, marked as answer]
Hi @marioaag here are a few links to our documentation about avocado varianters. I hope they would be useful to you.
https://avocado-framework.readthedocs.io/en/latest/guides/writer/chapters/parameters.html
https://avocado-framework.readthedocs.io/en/latest/guides/writer/chapters/parameters.html#varianter
https://avocado-framework.readthedocs.io/en/latest/guides/writer/chapters/parameters.html#varianter-plugins
https://avocado-framework.readthedocs.io/en/latest/plugins/optional/multiplexer.html
https://avocado-framework.readthedocs.io/en/latest/plugins/optional/varianter_cit.html
https://avocado-framework.readthedocs.io/en/latest/plugins/optional/varianter_pict.html
https://avocado-framework.readthedocs.io/en/latest/plugins/optional/varianter_yaml_to_mux.html#yaml-to-mux-plugin
